File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

#define INF 999999

int dr[] = {-1, 1, 0, 0};
int dc[] = {0, 0, -1, 1};

vector<vector<int>> board_ext;
int table[25][25][4]; // 상, 하, 좌, 우로부터 해당 좌표에 진입하는 경로의 최소 비용을 저장한다.
int N;

struct Node {
    int row, col;
    int dir; // 해당 노드에 방문하기 직전의 방향
    int cost;
    Node(int row, int col, int dir, int cost) {
        this->row = row;
        this->col = col;
        this->dir = dir;
        this->cost = cost;
    }
};

void bfs(int row, int col, int direction) {
    queue<Node> q;
    q.push(Node(row, col, direction, 0));
    
    while(!q.empty()) {
        Node cur = q.front();
        q.pop();
        
        int r = cur.row;
        int c = cur.col;
        int dir = cur.dir;
        int cost = cur.cost;
        
        for(int i=0; i<4; i++) {
            int nr = r + dr[i];
            int nc = c + dc[i];
            // 상하좌우 노드를 방문
            if (nr >=0 && nc >=0 && nr < N && nc < N && board_ext[nr][nc] == 0) {
                // 방향이 변하지 않았을 경우 +100, 변했을 경우 +600
                // 상->하, 좌->우 같은 형식으로 바뀌는 경우는 제외 (역행)
                int new_cost = (dir == i) ? (cost+100) : (cost+600);
                
                // 더 짧은 거리가 있을 경우 큐에 추가
                if (new_cost < table[nr][nc][i]) {
                    table[nr][nc][i] = new_cost;
                    q.push(Node(nr, nc, i, new_cost));
                }
                
            }
        }
        
    }
    
}

int solution(vector<vector<int>> board) {
    
    N = board.size();
    board_ext = board;
    
    // 1. dp 테이블을 초기화한다.
    for(int i=0; i<25; i++) {
        for(int j=0; j<25; j++) {
            for(int k=0; k<4; k++)
                table[i][j][k] = INF;
        }
    }
    
    // 2. bfs로 시작지점 -> 우, 시작지점 -> 하 2가지 경우를 계산한다.
    // bfs 안에서 걸러도 되는데, 조건문을 너무 많이 비교할 것 같음.
    bfs(0, 0, 3);
    bfs(0, 0, 1);
    
    int result = table[N-1][N-1][0];
    for(int i=0; i<4; i++) result = min(result, table[N-1][N-1][i]);
    
    return result;
}

File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.java
import java.util.*;

class Solution {
    
    static int[][] boardSt;
    static int N;
    static int[][][] table; // 각 진입 방향별로 최단 거리를 저장한다.
    static final int INF = 999999;
    
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};
    
    static class Node {
        int row, col;
        int dir; // 현재 좌표에 도달하기 직전의 진입 방향
        int cost;
        public Node(int row, int col, int dir, int cost) {
            this.row = row;
            this.col = col;
            this.dir = dir;
            this.cost = cost;
        }
    }
    
    static void bfs(int row, int col, int direction) {
        Queue<Node> q = new LinkedList<>();
        q.add(new Node(row, col, direction, 0));
        
        while(!q.isEmpty()) {
            Node cur = q.poll();
            
            int r = cur.row;
            int c = cur.col;
            int dir = cur.dir;
            int cost = cur.cost;
            
            for(int i=0; i<4; i++) {
                int nr = r + dr[i];
                int nc = c + dc[i];
                if (nr >=0 && nc >=0 && nr < N && nc < N && boardSt[nr][nc] == 0) {
                    // dir와 i가 같으면 진행 방향이 같다는 의미
                    // 직전 경로와 같으면 +100, 다르면 +600
                    // 상-하, 좌-우 변경은 되돌아가는 움직임인데, 거리 +600으로 처리되므로 사실상 업데이트 안 됨
                    int new_cost = (dir == i) ? cost+100 : cost+600;
                    if (new_cost < table[nr][nc][i]) {
                        table[nr][nc][i] = new_cost;
                        q.add(new Node(nr, nc, i, new_cost));
                    }
                }
            }
            
        }
        
    }
    
    public int solution(int[][] board) {
        N = board.length;
        boardSt = board;
        table = new int[N][N][4];
        
        // 1. dp 테이블을 초기화한다.
        for(int i=0; i<N; i++) {
            for(int j=0; j<N; j++) {
                for(int k=0; k<4; k++)
                    table[i][j][k] = INF;
            }
        }
        
        for(int i=0; i<4; i++) table[0][0][i] = 0;
        
        // 2. bfs를 돌려 최단 거리를 찾는다.
        bfs(0, 0, 1);
        bfs(0, 0, 3);
        
        int result = table[N-1][N-1][0];
        for(int i=0; i<4; i++) result = Math.min(result, table[N-1][N-1][i]);
        
        return result;
    }
}